### ISSUE TITLE
Php-CS-Fixer don't fix spaces after If

### ISSUE THREAD

**Issue by ShockedPlot7560** [pmmp/PocketMine-MP#5804]
### Issue description

Currently, if we add IFs in pocketmine directly with a space between the end of the condition and the open bracket, it won't be corrected by the CS fix.

```php
if (1 === 1) {
     // hello world
}
```

The above code will remain as it is, rather than being modified by: 
```php
if (1 === 1){
     // hello world
}
```

### OS and versions

* PocketMine-MP: 5.0.2+dev
* OS: Windows / Ubuntu 20.04

**Comment by dktapps** [comment]
Yeah, this is a known issue. Last time I looked, `braces` fixer wasn't granular enough to permit fixing this, since it made assumptions about PSR code style. Maybe this has changed since then.

**Comment by ShockedPlot7560** [comment]
And a pocketmine custom rule? Would that be possible?

**Comment by dktapps** [comment]
Never seemed worth the effort. You're welcome to investigate if you have the time, but I'd advise looking at the rule lists to see if there's anything that fills this need that we're not using. https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/list.rst

**Comment by dktapps** [comment]
For posterity, I'll paraphrase the Discord discussion on the topic of changing the CS to fit PHP-CS-Fixer here:

A lot of these CS problems would be avoided if we adopted a standard code-style as per PSR. However, I have several reasons to dislike this idea:
- All of the standard CS options in PSR involve ugly stuff such as opening braces on newlines, spaces for indentation etc. which I don't personally believe provides any value for a reader (wasting vertical screen space etc.)
- Changing PM to use a vanilla PSR standard code style would involve reformatting almost every line of code in the entire project due to the use of tabs for indentation. This is obviously not going to happen.
- Even if we don't go directly for a vanilla PSR change and just yield on what PHP-CS-Fixer can't give us (e.g. no spaces before braces), it will still cause thousands of lines of useless code changes that no one asked for.
- Any CS change on this scale would cause an enormous amount of chaos and pain for forks merging upstream changes with no practical benefit whatsoever.

We can of course write custom extensions to implement parts of the code style that PHP-CS-Fixer is incapable of, but this increases maintenance burden, since we would then need to worry about PHP-CS-Fixer API compatibility of our custom rules. I'm quite disappointed to see a [PHP-CS-Fixer collaborator's opinionated take](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7036#issuecomment-1579517398) on this (seriously? can it really be harder than adding a fixer option `num_spaces`???), as I don't think a CS tool should be limited to standards - after all, we're able to maintain current code style using PhpStorm, so I don't see why it shouldn't be possible with a third-party fixer.

It may be possible to do this using an alternative CS tool such as [phpcbf](https://github.com/squizlabs/PHP_CodeSniffer), but this would require a lot of time and effort to build up a codestyle configuration. I haven't explored this much yet, but I do know that [PHPStan](https://github.com/phpstan/build-cs) uses it, so it may be worth starting there.

**Comment by Wirone** [comment]
Hello everyone, I'm the PHP-CS-Fixer maintainer mentioned above, and I would like to make things clear 🙂.

**Fixer promotes standards** (PSR-2, PSR-12, now PER-CS), but it does not mean you can't make your own standard. You can create it by extending existing rule set, or by creating configuration from scratch, using any rules you want. If anything can't be done with existing, built-in rules, there is always possibility of creating [custom fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/custom_rules.rst?rgh-link-date=2023-06-06T22%3A07%3A56Z), registering it in the config and including it in the rules' configuration.

Having said that, I would like to clearly point out that **Fixer's Core Team has very limited capacity** and it's just impossible to implement any request reported by the users. It's also not a good idea to over-pack core with rules that cover edge cases, that are not widely used. Every rule has maintenance cost, so having in mind that currently we have only 2 active core members (me and @kubawerlos) + few contributors that not necessarily want to implement other people's ideas but rather prefer working on things they come up with, we try to limit amount of work as much as possible. In recent weeks I closed several issues (outdated, already fixed etc.), reviewed and merged multiple PRs (even resurrected one from 2018 IIRC), aiming for "0 backlog". We want to resolve as many issues as possible, so we can then focus on actual tool development and refactoring.

There are several big things to do: [parallel run](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/2803), introducing proper DI, introducing services (analysers) for token traversal to make rule development easier (currently a lot of logic is redundant between fixers) and so on... What made you think that any Core Team member would prefer working on superfluous space before `{` just to satisfy your needs 🤔?

> seriously? can it really be harder than adding a fixer option `num_spaces`???

**If it's that easy, I highly recommend creating a PR with a solution** 🤷‍♂️. Really, I don't understand your attitude here. As a user of several tools, I also create issues when I encounter any problem or I would like to see some feature. But at the same time I understand and fully respect each maintainer's approach and I don't complain behind the back if they close my request. I totally understand that it's not always possible to create a PR because it requires more knowledge and more time to do, but then - if you publicly state that something is easy, then just do it.

> We can of course write custom extensions to implement parts of the code style that PHP-CS-Fixer is incapable of, but this increases maintenance burden (...)

<img width="619" alt="image" src="https://github.com/pmmp/PocketMine-MP/assets/600668/6da61ffb-774a-444c-af9b-7daa6a1df810">

Seems legit.

> Yeah, this is a known issue. Last time I looked, braces fixer wasn't granular enough to permit fixing this, since it made assumptions about PSR code style. Maybe this has changed since then.

[`braces` fixer is deprecated](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/4885), please use [replacement rules](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules/basic/braces.rst#this-rule-is-deprecated-and-will-be-removed-on-next-major-version) instead.

To be honest I did not look at current rules and verify if your case is possible to cover, because I believe it's _your_ work to do. I just let you know that we won't work on it. Maybe I should add that if you come up with sane solution (as a PR), we can review it and maybe even merge it (no promises 😉).

PS. Maybe I shouldn't do it as a Fixer member, but there's also [Easy Coding Standard](https://github.com/easy-coding-standard/easy-coding-standard) that combines PHP-CS-Fixer's fixers with PHPCodeSniffer's sniffs, so you can define your standard using rules from both worlds. And it also includes several custom fixers too.

Finger's crossed for you and your own standard! Cheers 🍻

**Comment by kubawerlos** [comment]
Great explanation @Wirone.

One more thing:

> we would then need to worry about PHP-CS-Fixer API compatibility of our custom rules

Not much to worry about, all you have to do is implement [`FixerInterface`](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.17.0/src/Fixer/FixerInterface.php).

**Comment by dktapps** [comment]
>  I totally understand that it's not always possible to create a PR because it requires more knowledge and more time to do, but then - if you publicly state that something is easy, then just do it.

Yeah, I think that's fair enough. Our team was already exploring custom rules anyway.

The impression I got from your response to the issue was "we won't add this because it's not PSR and we don't like the idea", which struck me as weirdly opinionated and inconsistent with the existing PHP-CS-Fixer ruleset's available configuration options in many cases.

I hadn't yet looked at the codebase, but I know that several other rules including `concat_space` and `cast_spacing` support options for this (which PM already uses), so I thought it was weird to shut down the request on the basis of PSR and maintainability. These options already do what we needed for other rules, so I figured they can't be that difficult to do if non-standards-compliant options for similar spacing concerns are already scattered throughout other rules. Also, `curly_braces_position` itself already supports `same_line` placement, which isn't PSR-compliant.

I understand (and fully relate to) your point about asking someone else to maintain something we need (I have this exact same discussion with my users every other day), but this (to me) seems like a small improvement, and I can't see a reason why it would cause any meaningful maintenance cost (at least, no worse than is already invoked by other rules, unless there is some problem that I'm not seeing specifically for braces).

**Comment by dktapps** [comment]
For posterity (not related to PHP-CS-Fixer):

I would be OK with changing the spaces around braces specifically if it weren't for the avalanche of merge conflicts that would come about from such a change, both within the PM repo itself and everyone who has forked the repository (not to mention putting a big break in the commit history for stuff like `git blame`). I think the time for making sweeping code style changes like that to PM has long since passed by.

The more important thing we need to think about at this point is more about consistency within the project rather than aesthetic appeal. The current code style may not be standards compliant, but it is at least consistently used, so we only need to be concerned with enforcing it.

**Comment by Wirone** [comment]
@dktapps I think you may be interested in [`.git-blame-ignore-revs`](https://twitter.com/_Codito_/status/1660564114911920128) file in the repo - you can change the coding style, apply it to whole code base and then tell Git/Github to ignore such commit when calculating blame. Personally I wouldn't care about merge conflicts, since CS is automated everyone interested can just apply styling with a tool.

**Comment by dktapps** [comment]
> @dktapps I think you may be interested in [`.git-blame-ignore-revs`](https://twitter.com/_Codito_/status/1660564114911920128) file in the repo - you can change the coding style, apply it to whole code base and then tell Git/Github to ignore such commit when calculating blame. Personally I wouldn't care about merge conflicts, since CS is automated everyone interested can just apply styling with a tool.

That's cool, I didn't know such a feature existed. I'll look into that, thanks for the tip!
